ls1intum/Artemis
# `Programming exercises`: Synchronize repository locks after exercise update

### Checklist
#### General


#### Server

### Motivation and Context


Sometimes instructors have to change certain settings for programming exercises during their conduction. Some of these changes might need locking or unlocking the student repositories. This is now implemented.

### Description


The repositories should be **unlocked** if:
- the due date is postponed and was previously before now
- the offline IDE gets allowed or 
- the exercise gets released by changing the release date

They should be **locked** if:
- the due date is set in the past and was previously not
- if the offline IDE gets forbidden or
- when the exercise gets unreleased by changing the release date

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Programming Exercise

1. Log in to Artemis
2. Participate as a student
3. Do one of the mentioned changes and check if the repository got locked/unlocked correctly (Go to the repository settings in BitBucket as instructor and check the access rights of the student). The changes might not happen instantly since they have to be broadcasted through the InstanceMessegingService.
4. Check the other possible changes

#### Exam Mode Testing



Prerequisites:
- 1 Instructor
- 1 Students
- 1 Exam with a Programming Exercise

1. Log in to Artemis
2. Participate in the exam as a student
3. Check that allowing/forbidding the offline IDE works here as well

### Review Progress






#### Code Review
#### Manual Tests
#### Exam Mode Test

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| ProgrammingExerciseService.java | 82% | 92% |
